src/shims/unix/fd.rs

```diff
@@ -310,6 +382,45 @@ impl FdTable {
 
 impl<'tcx> EvalContextExt<'tcx> for crate::MiriInterpCx<'tcx> {}
 pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+    /// For a specific unique file descriptor id, get its ready events and update the corresponding ready lists
+    fn check_and_update_readiness(
+        &self,
+        id: FdId,
+        get_ready_events: impl FnOnce() -> InterpResult<'tcx, EpollReadyEvents>,
+    ) -> InterpResult<'tcx, ()> {
+        let this = self.eval_context_ref();
+        // Get a list of EpollEventInterest that is associated to a specific file description.
+        if let Some(epoll_interests) = this.machine.epoll_interests.get_epoll_interest(id) {
+            // Retrieve the readiness events of the file description.
+            let epoll_ready_events = get_ready_events()?;
+
+            // Get the bitmask of ready events.
+            let epollin = this.eval_libc_u32("EPOLLIN");
+            let epollout = this.eval_libc_u32("EPOLLOUT");
+            let epollrdhup = this.eval_libc_u32("EPOLLRDHUP");
+            let ready_events = epoll_ready_events.get_event_bitmask(epollin, epollout, epollrdhup);
+
+            for weak_epoll_interest in epoll_interests {
+                if let Some(epoll_interest) = weak_epoll_interest.upgrade() {
+                    // This checks if any of the events specified in epoll_event_interest.events
+                    // match those in ready_events.
+                    let epoll_event_interest = epoll_interest.borrow();
+                    let flags = epoll_event_interest.events & ready_events;
+                    // If there is any event that we are interested in being specified as ready,
+                    // insert an epoll_return to the ready list.
+                    if flags != 0 {
+                        let epoll_key = (id, epoll_event_interest.file_descriptor);
+                        let ready_list = &mut epoll_event_interest.ready_list.borrow_mut();
+                        let epoll_return =
+                            EpollEventInstance::new(flags, epoll_event_interest.data);
+                        ready_list.insert(epoll_key, epoll_return);
```

Comment:
this overwrites any previous flags that were in there, is that intended (add a test for whatever the expected behaviour is)